level2/main.tf

```
@@ -119,21 +119,76 @@ data "aws_ami" "amazon_linux" {
   }
 }
 
+resource "aws_iam_policy" "s3policy" {
+  name        = "s3_permissionPolicy"
+  description = "Permission Level to access S3"
+
+  policy = jsonencode({
+    "Version": "2012-10-17",
+    "Statement": [
+        {
+            "Effect": "Allow",
+            "Action": [
+                "s3:*",
+                "s3-object-lambda:*"
+            ],
+            "Resource": "*"
+        }
+    ]
+  })
+}
+
+resource "aws_iam_role" "ec2instrole" {
+  name = "EC2_InstanceRole"
+  path = "/"
+
+  assume_role_policy = <<EOF
+{
+    "Version": "2012-10-17",
+    "Statement": [
+        {
+            "Action": "sts:AssumeRole",
+            "Principal": {
+               "Service": "ec2.amazonaws.com"
+            },
+            "Effect": "Allow",
+            "Sid": ""
+        }
+    ]
+}
+EOF
+}
+
+resource "aws_iam_policy_attachment" "policy_role_attach" {
+  name       = "InstancePolicyRole"
+  roles      = [aws_iam_role.ec2instrole.name]
+  policy_arn = aws_iam_policy.s3policy.arn
+}
+
+resource "aws_iam_instance_profile" "ec2instprofile" {
+  name = "EC2_instanceProfile"
```

```suggestion
  name = "EC2_InstanceRole"
```

Role name and instance profile name should match. Otherwise it could cause bugs